### PR TITLE
feat: add list of linters from npm

### LINKED ISSUE #2: List linters of npm
Please, add list linters of npm to project. 

List linters it's a file which will include: 
- name of manager
- list of linters 

I suggest the following structure for this file: 
```JSON
{
    "manager": "managerName",
    "linters": [
         "linter1",
         "linter2"
    ]
}
```

### PR DESCRIPTION
**Pull request by Arcle1**
- Add list of linters from npm

Closes #2

### CODE REVIEW

File: script/gulp/task.build.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+// Gulp build task
+const build = () => gulp
+    .src(config.src.linters)
+    .pipe(getLintersData())
+    .pipe(gulp.dest((file) => {
+        let content = JSON.parse(file.contents.toString('utf8'));
+        return path.join(config.build.dir, content.name);
+    }))
+    .pipe(gulp.dest((file) => {
+        let content = JSON.parse(file.contents.toString('utf8'));
```
Comment (svoboda-rabstvo) [inline script/gulp/task.build.js]: it's const ?)

File: script/gulp/task.build.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    return json;
+};
+
+/**
+ * Get array of objects, where each contains meta of linter and base & path where it will be stored
+ * @return {array} obj - array of objects with data of linters
+ */
+const getLintersData = () => {
+    return through2.obj(async function (file, enc, next) {
+        const lintersArray = core.fnc.bufferToJson(file.contents);
```
Comment (svoboda-rabstvo) [inline script/gulp/task.build.js]: I think we need better name for that

File: script/gulp/task.build.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ */
+const validateMeta = (meta) => {
+    if (!meta.name) throw RequiredFieldInvalidError;
+    if (!meta.version) throw RequiredFieldInvalidError;
+    return true;
+};
+
+const makeSchemaGreatArain = (meta) => {
+    let json = {};
+    json.$schema = 'https://schema.linterhub.com/package.json';
```
Comment (svoboda-rabstvo) [inline script/gulp/task.build.js]: We neeed to get `uri` from `@linterhub/schema` package, also `version` too

File: script/gulp/task.build.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    json.package = `${meta.name}:${meta.version}`;
+    return json;
+};
+
+/**
+ * Get array of objects, where each contains meta of linter and base & path where it will be stored
+ * @return {array} obj - array of objects with data of linters
+ */
+const getLintersData = () => {
+    return through2.obj(async function (file, enc, next) {
```
Comment (svoboda-rabstvo) [inline script/gulp/task.build.js]: I suggest to separate function inside the `through2.obj`

File: script/gulp/task.build.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+// External modules as aliases
+const core = require('./index.js');
+const gulp = core.amd.gulp;
+const registry = core.amd.registry;
+const path = core.amd.path;
+const through2 = core.amd.through2;
+const File = core.amd.file;
+
+// Create errors
+const RequiredFieldInvalidError = new Error('One or more of required fields in meta are invalid.');
```
Comment (svoboda-rabstvo) [inline script/gulp/task.build.js]: This function move out to `index.js`

File: script/gulp/task.build.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        next();
+    });
+};
+
+// Gulp build task
+const build = () => gulp
+    .src(config.src.linters)
+    .pipe(getLintersData())
+    .pipe(gulp.dest((file) => {
+        const content = core.fnc.bufferToJson(file.contents);
```
Comment (svoboda-rabstvo) [inline script/gulp/task.build.js]: I suggest to initialize all core functions in separate block beside the `External modules as aliases`

File: script/gulp/task.build.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+/**
+ * Get array of objects, where each contains meta of linter and base & path where it will be stored
+ * @return {array} obj - array of objects with data of linters
+ */
+const getLintersData = () => {
+    return through2.obj(async function (file, enc, next) {
+        const lintersArray = core.fnc.bufferToJson(file.contents);
+
+        const manager = registry.getManager(lintersArray.manager);
+        let metaArray = await Promise.all(
```
Comment (svoboda-rabstvo) [inline script/gulp/task.build.js]: I'm not sure about this name

File: script/gulp/task.build.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+            lintersArray.linters.map(async (lint) => {
+                let meta = await manager.getMeta(lint);
+                meta = Object.assign(makeSchemaGreatArain(meta), meta);
```
Comment (svoboda-rabstvo) [inline script/gulp/task.build.js]: `Object.assign` don't copy dependent property in low lvl

File: script/gulp/task.build.js [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+ * @param {array} meta - array with meta information of linter
+ * @return {boolean} - if validation passed without errors
+ */
+const validateMeta = (meta) => {
+    if (!meta.name) throw RequiredFieldInvalidError;
+    if (!meta.version) throw RequiredFieldInvalidError;
+    return true;
+};
+
+const makeSchemaGreatArain = (meta) => {
```
Comment (svoboda-rabstvo) [inline script/gulp/task.build.js]: Change name for this function please ;)

File: script/gulp/index.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    jsonToBuffer: (json) => Buffer.from(JSON.stringify(json), 'utf8'),
+    bufferToJson: (buffer) => JSON.parse(buffer.toString('utf8')),
+};
+
+// Exported shared config, modules and functions
+module.exports = {
+    cfg: cfg,
+    amd: amd,
+    fnc: fnc,
+};
```
Comment (svoboda-rabstvo) [inline script/gulp/index.js]: Blank line in end of file

File: script/gulp/task.build.js [hunk 11]
```diff
@@ -0,0 +1,80 @@
+'use strict';
+
+// Import config
+const config = require('./config.json');
+
+// External modules as aliases
+const core = require('./index.js');
```
Comment (svoboda-rabstvo) [inline script/gulp/task.build.js]: I think has better import `core` like a global   (like in `schema` repo)

File: script/gulp/task.build.js (hunk 8 above)
Comment (Arcle1) [reply script/gulp/task.build.js]: @svoboda-rabstvo it's just for 2 objects concatenation and place it in `meta`.

File: gulpfile.js [hunk 12]
```diff
@@ -0,0 +1,16 @@
+'use strict';
+
+// Set global shared core
+const lhcore = require('./script/gulp/index.js');
+global.lhcore = lhcore;
+
+// External modules as aliases
+const gulp = require('gulp');
+const hubRegistry = require('gulp-hub');
```
Comment (svoboda-rabstvo) [inline gulpfile.js]: I think better name is `gulpHub`

File: package.json [hunk 13]
```diff
@@ -0,0 +1,12 @@
+{
+    "scripts": {},
```
Comment (svoboda-rabstvo) [inline package.json]: Why `author`, `license` and e.t.c is not exist?

File: script/gulp/task.build.js [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    }
+};
+
+/**
+ * Get array of objects, where each contains meta of linter and base & path where it will be stored
+ * @return {array} obj - array of objects with data of linters
+ */
+const getMeta = () => {
+    return through2.obj(async function (file, enc, next) {
+        const linters = bufferToJson(file.contents);
```
Comment (svoboda-rabstvo) [inline script/gulp/task.build.js]: Mb `packages` is better name? @Arcle1 What do u think?

File: script/gulp/task.build.js [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+
+        const manager = registry.getManager(linters.manager);
+        let metaLinters = await Promise.all(
+            linters.linters.map(async (lint) => {
+                let meta = await manager.getMeta(lint);
+                meta = getPackageSchema(meta);
+                return validateMeta(meta) ? meta : '';
```
Comment (svoboda-rabstvo) [inline script/gulp/task.build.js]: I think `validate` function don't need, at least for now

File: script/gulp/task.build.js [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+ */
+const getMeta = () => {
+    return through2.obj(async function (file, enc, next) {
+        const linters = bufferToJson(file.contents);
+
+        const manager = registry.getManager(linters.manager);
+        let metaLinters = await Promise.all(
+            linters.linters.map(async (lint) => {
+                let meta = await manager.getMeta(lint);
+                meta = getPackageSchema(meta);
```
Comment (svoboda-rabstvo) [inline script/gulp/task.build.js]: It's `package.json`, not `meta` :)

File: gulpfile.js (hunk 12 above)
Comment (Arcle1) [reply gulpfile.js]: @svoboda-rabstvo we have hubRegistry in `schema`. Do you think we should rename it here?

File: script/gulp/task.build.js (hunk 14 above)
Comment (Arcle1) [reply script/gulp/task.build.js]: @svoboda-rabstvo ok.

File: script/gulp/task.build.js (hunk 15 above)
Comment (Arcle1) [reply script/gulp/task.build.js]: @svoboda-rabstvo ok.

File: script/gulp/task.build.js (hunk 16 above)
Comment (Arcle1) [reply script/gulp/task.build.js]: @svoboda-rabstvo yes, it is package.json.
We send object `meta` to it, but get properties from `schema.package`.

```
$schema: schema.package.$id,
$version: schema.package.$version,
```

File: gulpfile.js (hunk 12 above)
Comment (svoboda-rabstvo) [reply gulpfile.js]: Yes, because package named is `gulp-hub`, not `hubRegistry`

File: script/gulp/task.build.js (hunk 16 above)
Comment (svoboda-rabstvo) [reply script/gulp/task.build.js]: This function return `package.json` schema, not `meta` from registry, right?

File: script/gulp/task.build.js [hunk 17]
```diff
@@ -0,0 +1,92 @@
+'use strict';
+
+// Import config
+const config = require('./config.json');
```
Comment (svoboda-rabstvo) [inline script/gulp/task.build.js]: We can use `core` for import config. `require` is don't need

File: gulpfile.js (hunk 12 above)
Comment (Arcle1) [reply gulpfile.js]: @svoboda-rabstvo ok.

File: script/gulp/task.build.js (hunk 16 above)
Comment (Arcle1) [reply script/gulp/task.build.js]: @svoboda-rabstvo yes, you are right!

File: script/gulp/index.js [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+
+// Shared functions
+const fnc = {
+    readJson: (path) => JSON.parse(amd.fs.readFileSync(path)),
+    jsonToBuffer: (json) => Buffer.from(JSON.stringify(json), 'utf8'),
+    bufferToJson: (buffer) => JSON.parse(buffer.toString('utf8')),
+};
+
+// Shared errors
+const err = {
```
Comment (svoboda-rabstvo) [inline script/gulp/index.js]: It's not used now ;)

File: package.json [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+    "description": "Catalog of linters",
+    "homepage": "",
+    "license": "MIT",
+    "contributors": [],
+    "keywords": [
+        "linterhub"
+    ],
+    "repository": "linterhub/catalog",
+    "bugs": "https://github.com/linterhub/catalog/issues",
+    "main": "index.js",
```
Comment (svoboda-rabstvo) [inline package.json]: We don't have `index.js`. Please, remove it

**Pull request merged**